src/classes/RD2_ApiService_TEST.cls

```diff
@@ -246,6 +247,116 @@ private class RD2_ApiService_TEST {
         System.assertEquals(1, isActiveCurrentNotPaused, 'There should be one active and current unpaused schedules returned');
     }
 
+    /**
+     * @description Validate an empty installment response when RD2 is not enabled
+     */
+    @IsTest
+    private static void shouldReturnNoInstallmentsWhenRD2NotEnabled() {
+        npe03__Recurring_Donation__c rd = rdGateway.getRecords()[0];
+
+        Map<Id, List<Installment>> installmentsByRd = retrieveInstallmentsUsingApi(
+            date.today(),
+            date.today().addYears(2),
+            rd.Id);
+
+        System.assertEquals(0, installmentsByRd.size(),
+            'There should be no recurring donation installments returned when RD2 is not enabled');
+    }
+
+    /**
+     * @description Validate an empty installment response when invalid RD Id is passed in
+     */
+    @IsTest
+    private static void shouldReturnNoInstallmentsWithInvalidRDId() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Map<Id, List<Installment>> installmentsByRd = retrieveInstallmentsUsingApi(
+            date.today(),
+            date.today().addYears(2),
+            UTIL_UnitTestData_TEST.mockId(npe03__Recurring_Donation__c.SObjectType));
+
+        System.assertEquals(0, installmentsByRd.size(),
+            'There should be no recurring donation installments returned when RD Id is invalid');
+    }
+
+    /**
+     * @description Validate an empty installment response when empty RD Id set is passed in
+     */
+    @IsTest
+    private static void shouldReturnNoInstallmentsWithEmptyRDIdSet() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Map<Id, List<Installment>> installmentsByRd = retrieveInstallmentsUsingApi(
+            Date.today(),
+            Date.today().addYears(2),
+            null);
+
+        System.assertEquals(0, installmentsByRd.size(),
+            'There should be no recurring donation installments returned when RD Id Set is empty');
+    }
+
+    /**
+     * @description Validate a valid installment response when start date is not passed in
+     */
+    @IsTest
+    private static void shouldReturnValidInstallmentsWithMissingStartDate() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+        npe03__Recurring_Donation__c rd = rdGateway.getRecords()[0];
+
+        Map<Id, List<Installment>> installmentsByRd = retrieveInstallmentsUsingApi(
+            null,
+            date.today().addYears(2),
+            rd.Id);
+
+        System.assertNotEquals(0, installmentsByRd.size(),
+            'Recurring donation installments should be returned when Start Date is missing');
+    }
+
+    /**
+     * @description Validate an empty installment response when End Date is missing
+     */
+    @IsTest
+    private static void shouldReturnNoInstallmentsWithMissingEndDate() {
```

Comment:
Why not make both start and end date as optional, and if the end date is not specified, then return default 12 installments?
However, the End Date is required based on the required param check in the `Callable_API`, so I am wondering why no exception is thrown when the end date is not provided?

Reply:
I don't believe a default end date makes sense.  We could be dealing with any period - weekly, monthly, yearly - so 12 installments could represent a wide range of dates.  Since the purpose of this API is to assist in custom revenue forecasting scenarios, a specific end date makes sense to me.

Reply:
Pretty sure that the reason no exception is thrown in this test currently is that an end date parameter is being provided.  It is null, but it exists.  Exception would be thrown when a required parameter is not provided at all.